File: api/src/device.go
package main

import (
	"fmt"
	"net/http"
	"os"
	"path/filepath"
)

// Reusable function to parse and server the JSON we loaded into memory, and fall back to on-disk JSON if required.
func serveDevice(writer http.ResponseWriter, request *http.Request, mapping string) {
	id, subkeys, ok := extractPathAfter(mapping, request.URL.Path)
	if !ok {
		// Error: extractPathAfter failed, bad request
		http.Error(writer, badRequest, http.StatusBadRequest)
		fmt.Printf("[Request] Unable to parse request \"%s\"\n", request.URL.RequestURI())
		return
	}

	var dict map[string]interface{}

	switch mapping {
	case "boardconfig": dict = boardconfigDevices
	case "identifier": dict = identifierDevices
	case "model": dict = modelDevices
	}

	lookup, ok := dict[id] 
	if !ok {
		fmt.Printf("[Request] Couldn't find that device in memory, restarting from on-disk.\n")
		lookupMap, ok := mappings[mapping].(map[string]interface{})
		if !ok {
			http.Error(writer, "", http.StatusInternalServerError)
			return
		}

		val, exists := lookupMap[id]
		if !exists {
			// Error: Mapping didn't return data for that device
			http.Error(writer, badDevice, http.StatusNotFound)
			fmt.Printf("[Request] Unable to find device \"%s\"\n", id)
			return
		}

		valStr, ok := val.(string)
		if !ok {
			// Error: Unable to convert JSON path into string
			http.Error(writer, badDataRead, http.StatusInternalServerError)
			fmt.Printf("[Request] Unable to convert value to string \"%s\"\n", val)
			return
		}

		filePath := filepath.Join(valStr)

		fmt.Printf("[Request] Getting ready to serve the %s data from \"%s\"\n", mapping, filePath)

		if _, err := os.Stat(filePath); os.IsNotExist(err) {
			// Error: Device file is missing
			http.Error(writer, badPath, http.StatusInternalServerError)
			fmt.Printf("[Request] Unable to locate \"%s\"\n", filePath)
			return
		}

		fmt.Printf("[Request] Found our file at \"%s\", now parsing it.\n", filePath)

		var err error
		lookup, err = loadJSON(filePath)
		if err != nil {
			// Error: Device file is unloadable
			http.Error(writer, badDataRead, http.StatusInternalServerError)
			fmt.Printf("[Request] Unable to parse the JSON at \"%s\"\n", filePath)
			return
		}
	}

	jsonResponse := parseDeviceJSON(writer, request, subkeys, lookup)
	if jsonResponse == nil { return }

	if !respondToGet(writer, jsonResponse) {
		http.Error(writer, badResponse, http.StatusInternalServerError)
		fmt.Printf("[Request] Failed to send a response to the client.\n")
		return
	}
}

File: api/src/Utilities/DeviceUtils/Device.swift
//
//  Utilities+Device.swift
//  Labradorite-Server
//
//  Created by Eva Isabella Luna on 10/18/25.
//

import Foundation
import Network

class Device {
	var memory: Memory!
	
	init() {
		self.memory = Memory(parent: self)
	}
	
	func serveDevice(connection: NWConnection, method: String, path: String, headers: [String: String]) {
		let usedTarget: String
		// determine which target this call comes from by path
		if path.localizedCaseInsensitiveContains("/identifier/") { usedTarget = "identifier" }
		else if path.localizedCaseInsensitiveContains("/model/") { usedTarget = "model" }
		else if path.localizedCaseInsensitiveContains("/boardconfig/") { usedTarget = "boardconfig" }
		else { // fallback try default extraction of first mapping found in path
			usedTarget = ["boardconfig","model","identifier"].first(where: { path.localizedCaseInsensitiveContains("/\($0)/") }) ?? "identifier"
		}
		
		let (id, subkeys, ok) = utilities.extractPathAfter(target: usedTarget, path: path)
		guard ok else {
			utilities.log("[Request] Unable to parse request \"%@\"", path)
			let body = Data(badRequest.utf8)
			utilities.http.respondHeadersAndBody(connection: connection, status: 400, body: body)
			return
		}
		
		var dict: [String: Any]
		switch usedTarget {
		case "boardconfig": dict = boardconfigDevices
		case "model": dict = modelDevices
		default: dict = identifierDevices
		}
		
		if let lookup = dict[id] {
			// found in memory
			if let response = utilities.json.parseDeviceJSON(subkeys: subkeys, data: lookup, headers: headers) {
				utilities.http.respondHeadersAndBody(connection: connection, status: 200, body: response)
				return
			} else {
				let body = Data(badKey.utf8)
				utilities.http.respondHeadersAndBody(connection: connection, status: 400, body: body)
				return
			}
		}
		
		utilities.log("[Request] Couldn't find that device in memory, restarting from on-disk.")
		guard let lookupMap = mappings[usedTarget] else {
			let body = Data("".utf8)
			utilities.http.respondHeadersAndBody(connection: connection, status: 500, body: body)
			return
		}
		
		guard let mappedPathRaw = lookupMap[id] else {
			utilities.log("[Request] Unable to find device \"%@\"", id)
			let body = Data(badDevice.utf8)
			utilities.http.respondHeadersAndBody(connection: connection, status: 404, body: body)
			return
		}
		
		guard let mappedPath = mappedPathRaw as? String else {
			utilities.log("[Request] Unable to convert value to string \"%@\"", String(describing: mappedPathRaw))
			let body = Data(badDataRead.utf8)
			utilities.http.respondHeadersAndBody(connection: connection, status: 500, body: body)
			return
		}
		
		let filePath = mappedPath
		utilities.log("[Request] Getting ready to serve the %@ data from \"%@\"", usedTarget, filePath)
		
		if !FileManager.default.fileExists(atPath: filePath) {
			utilities.log("[Request] Unable to locate \"%@\"", filePath)
			let body = Data(badPath.utf8)
			utilities.http.respondHeadersAndBody(connection: connection, status: 500, body: body)
			return
		}
		
		utilities.log("[Request] Found our file at \"%@\", now parsing it.", filePath)
		do {
			let loaded = try utilities.json.loadJSON(at: filePath)
			if let response = utilities.json.parseDeviceJSON(subkeys: subkeys, data: loaded, headers: headers) {
				utilities.http.respondHeadersAndBody(connection: connection, status: 200, body: response)
				return
			} else {
				let body = Data(badKey.utf8)
				utilities.http.respondHeadersAndBody(connection: connection, status: 400, body: body)
				return
			}
		} catch {
			utilities.log("[Request] Unable to parse the JSON at \"%@\"", filePath)
			let body = Data(badDataRead.utf8)
			utilities.http.respondHeadersAndBody(connection: connection, status: 500, body: body)
			return
		}
	}
	
}
